pkg/controllers/cloud/virtualmachine_controller.go

```diff
@@ -33,95 +38,174 @@ const (
-		Scheme: r.Scheme,
+		Ch:     make(chan watch.Event, converterChannelBuffer),
+		Scheme: r.scheme,
+	}
+	return r
+}
+
+// Start performs the initialization of the controller.
+// VM controller is said to be initialized and synced when all the VM and EE
+// CRs are reconciled.
+func (r *VirtualMachineManager) Start() error {
+	if err := GetControllerSyncStatusInstance().waitForControllersToSync(
+		[]controllerType{ControllerTypeEE, ControllerTypeCES, ControllerTypeCPA}, syncTimeout); err != nil {
+		return err
 	}
-	if err := ctrl.NewControllerManagedBy(mgr).For(&cloudv1alpha1.VirtualMachine{}).Complete(r); err != nil {
+	r.log.Info("Init done", "controller", ControllerTypeVM.String())
+
+	vmMap := r.getAllVMs()
+	if err := r.syncExternalEntities(vmMap); err != nil {
 		return err
 	}
+	if err := r.syncExternalNodes(vmMap); err != nil {
+		return err
+	}
+
+	// Start the converter module.
 	go r.converter.Start()
-	return mgr.Add(r)
+	// Set sync status as complete.
+	GetControllerSyncStatusInstance().SetControllerSyncStatus(ControllerTypeVM)
+	// Setup watcher after VM cache is populated and EE CRs are reconciled.
+	if err := r.resetWatcher(); err != nil {
+		return err
+	}
+	// Blocking thread to wait for any VM event.
+	if err := r.processEvent(); err != nil {
+		return err
+	}
+	return nil
 }
 
-// Start performs the initialization of the controller.
-// A controller is said to be initialized only when the dependent controllers
-// are synced, and it keeps a count of pending CRs to be reconciled.
-func (r *VirtualMachineReconciler) Start(context.Context) error {
-	if err := GetControllerSyncStatusInstance().waitForControllersToSync([]controllerType{ControllerTypeEE}, syncTimeout); err != nil {
-		r.Log.Error(err, "dependent controller sync failed", "controller", ControllerTypeEE.String())
+// getAllVMs returns a map which contains all the VMs.
+func (r *VirtualMachineManager) getAllVMs() map[types.NamespacedName]struct{} {
+	vmObjList := r.Inventory.GetAllVms()
+	tempVmMap := map[types.NamespacedName]struct{}{}
+	for _, vmObj := range vmObjList {
+		vm := vmObj.(*runtimev1alpha1.VirtualMachine)
+		vmNn := types.NamespacedName{
+			Name:      vm.Name,
+			Namespace: vm.Namespace,
+		}
+		tempVmMap[vmNn] = struct{}{}
+	}
+	return tempVmMap
+}
+
+// syncExternalEntities validates that each EE has corresponding VM. If it does not exist then
+// the EE will be deleted.
+func (r *VirtualMachineManager) syncExternalEntities(vmMap map[types.NamespacedName]struct{}) error {
+	eeList := &antreav1alpha2.ExternalEntityList{}
+	if err := r.Client.List(context.TODO(), eeList, &client.ListOptions{}); err != nil {
 		return err
 	}
-	vmList := &cloudv1alpha1.VirtualMachineList{}
-	if err := r.Client.List(context.TODO(), vmList, &client.ListOptions{}); err != nil {
+	for _, ee := range eeList.Items {
+		eeLabelKeyName, exists := ee.Labels[config.ExternalEntityLabelKeyName]
+		if !exists {
```

Comment:
may be you can use external NodeName to filter none EE created by nephe

Reply:
Done